From dda5a5f2b5f753dcbc50efb4aa4a7c3ebf1a52d6 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Sun, 26 Jun 2016 19:43:57 +0300
Subject: [PATCH 533/538] pp3: graceful wait for net-if stop

When interface goes over stop/down there still could be
some packets "txdone_todo".
Sleep for (MV_PP3_TXDONE_TIMER_USEC_PERIOD+1jiffies),
waiting for TXDONE_TIMER releases everything by it's run
before the txdone_is_empty() checking.

Change-Id: I76143dcde57df4fd456fa61709a745dc3d41fcd8
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30734
Reviewed-by: Dovrat Zifroni <dovrat@marvell.com>
Tested-by: Dovrat Zifroni <dovrat@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Tested-by: Dmitri Epshtein <dima@marvell.com>
---
 drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
index 2920f65..b0dbb819 100644
--- a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
+++ b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
@@ -2858,14 +2858,14 @@ int mv_pp3_dev_stop(struct net_device *dev)
 		mv_pp3_gop_port_events_mask(emac);
 		/* disable EMAC RX processing */
 		mv_pp3_emac_rx_enable(emac, false);
-
-		mv_pp3_dev_down(dev_priv);
 	}
+	/* Down IF_LINK_UP */
+	mv_pp3_dev_down(dev_priv);
 
 	/* set device state in FW to disable */
 	mv_pp3_dev_fw_down(dev_priv);
 
-	mdelay(10);
+	msleep(MV_PP3_TXDONE_TIMER_USEC_PERIOD/1000 + jiffies_to_msecs(1));
 
 	mv_pp3_dev_napi_disable(dev_priv);
 
@@ -2876,6 +2876,9 @@ int mv_pp3_dev_stop(struct net_device *dev)
 	if (mv_pp3_dev_queues_proc_done(dev_priv))
 		return -1;
 
+	/* Wait for TXDONE timer passed with releasing evrything needed */
+	msleep(MV_PP3_TXDONE_TIMER_USEC_PERIOD/1000 + jiffies_to_msecs(1));
+
 	mv_pp3_dev_pools_empty(dev_priv);
 
 	if (!mv_pp3_dev_txdone_is_empty(dev_priv)) {
-- 
1.9.1

